packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1884 @@
+ */
+export type PermissionControllerState<Permission> =
+  Permission extends GenericPermission
+    ? {
+        subjects: PermissionControllerSubjects<Permission>;
+      }
+    : never;
+
+/**
+ * The state metadata of the {@link PermissionController}.
+ */
+const stateMetadata: StateMetadata<PermissionControllerState<any>> = {
+  subjects: { persist: true, anonymous: true },
+};
+
+// Typecast: We can't reference the class generics, so we must resort to "any"
+/**
+ * The default state of the {@link PermissionController}.
+ */
+const defaultState: PermissionControllerState<any> = {
+  subjects: {},
+};
+
+/**
+ * Gets the state of the {@link PermissionController}.
+ */
+export type GetPermissionControllerState = {
+  type: `${typeof controllerName}:getState`;
+  handler: () => PermissionControllerState<GenericPermission>;
+};
+
+/**
+ * Gets the names of all subjects from the {@link PermissionController}.
+ */
+export type GetSubjects = {
+  type: `${typeof controllerName}:getSubjectNames`;
+  handler: () => (keyof PermissionControllerSubjects<GenericPermission>)[];
+};
+
+/**
+ * Checks whether the specified subject has any permissions.
+ */
+export type HasPermissions = {
+  type: `${typeof controllerName}:hasPermissions`;
+  handler: GenericPermissionController['hasPermissions'];
+};
+
+/**
+ * Clears all permissions from the {@link PermissionController}.
+ */
+export type ClearPermissions = {
+  type: `${typeof controllerName}:clearPermissions`;
+  handler: () => void;
+};
+
+// TODO: Implement all desired actions
+/**
+ * The {@link ControllerMessenger} actions of the {@link PermissionController}.
+ */
+export type PermissionControllerActions =
+  | ClearPermissions
+  | GetPermissionControllerState
+  | GetSubjects
+  | HasPermissions;
+
+/**
+ * The generic state change event of the {@link PermissionController}.
+ */
+export type PermissionControllerStateChange = {
+  type: `${typeof controllerName}:stateChange`;
+  payload: [PermissionControllerState<GenericPermission>, Patch[]];
+};
+
+/**
+ * The {@link ControllerMessenger} events of the {@link PermissionController}.
+ *
+ * The permission controller only emits its generic state change events.
+ * Consumers should use selector subscriptions to subscribe to relevant
+ * substate.
+ */
+export type PermissionControllerEvents = PermissionControllerStateChange;
+
+/**
+ * The external {@link ControllerMessenger} actions available to the
+ * {@link PermissionController}.
+ */
+type AllowedActions =
+  | AddApprovalRequest
+  | HasApprovalRequest
+  | AcceptApprovalRequest
+  | RejectApprovalRequest;
+
+/**
+ * The messenger of the {@link PermissionController}.
+ */
+export type PermissionControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  PermissionControllerActions | AllowedActions,
+  PermissionControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+export type GenericPermissionController = PermissionController<
```

Comment:
Can we resolve this conversation, or do we need to take any action?

Reply:
I think we could get rid of it. Not a blocker, but this seems like a bad pattern in general, and it's avoidable in this case.

We use this type right now for `hasPermissions` and the permission middleware. For the middleware, we can pass through generic parameters and use the _real_ `PermissionController` type. For `hasPermissions`, it has a simple type that we can manually type out (`(origin: OriginString) => boolean`).

If we're in this situation in the future, we can consider adding generic parameters to our exported actions and/or events. I don't think we need to do that here, but it might be necessary someday to preserve type safety for actions/events.

Reply:
Hah, I actually access `_executeRestrictedMethod` on a `PermissionController` generic parameter. Accessing private / protected properties on type parameters is apparently expressly forbidden, but fine on the generic version. It seems fine as-is to me, and I'll just leave it for now if we can agree it's not a blocker!